operator/cmd/certificates/watcher.go

```diff
@@ -0,0 +1,137 @@
+package certificates
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// TODO: refactor code below to be testable and also tested
+const (
+	certificateRenewalInterval = 6 * time.Hour
+	ServerKey                  = "tls.key"
+	ServerCert                 = "tls.crt"
+)
+
+type CertificateWatcher struct {
+	apiReader             client.Reader
+	fs                    afero.Fs
+	certificateDirectory  string
+	namespace             string
+	certificateSecretName string
+	Log                   logr.Logger
+}
+
+func NewCertificateWatcher(mgr manager.Manager, namespace string, secretName string, log logr.Logger) *CertificateWatcher {
+	return &CertificateWatcher{
+		apiReader:             mgr.GetAPIReader(),
+		fs:                    afero.NewOsFs(),
+		certificateDirectory:  mgr.GetWebhookServer().CertDir,
+		namespace:             namespace,
+		certificateSecretName: secretName,
+		Log:                   log,
+	}
+}
+
+func (watcher *CertificateWatcher) watchForCertificatesSecret() {
+	for {
+		<-time.After(certificateRenewalInterval)
+		watcher.Log.Info("checking for new certificates")
+		if updated, err := watcher.updateCertificatesFromSecret(); err != nil {
+			watcher.Log.Info("failed to update certificates", "error", err)
+		} else if updated {
+			watcher.Log.Info("updated certificate successfully")
+		}
+	}
+}
+
+func (watcher *CertificateWatcher) updateCertificatesFromSecret() (bool, error) {
+	var secret corev1.Secret
+
+	err := watcher.apiReader.Get(context.TODO(),
+		client.ObjectKey{Name: watcher.certificateSecretName, Namespace: watcher.namespace}, &secret)
+	if err != nil {
+		return false, err
+	}
+
+	watcher.Log.Info("checking dir", "watcher.certificateDirectory ", watcher.certificateDirectory)
+	if _, err = watcher.fs.Stat(watcher.certificateDirectory); os.IsNotExist(err) {
+		err = watcher.fs.MkdirAll(watcher.certificateDirectory, 0755)
+		if err != nil {
+			return false, fmt.Errorf("could not create cert directory: %s", err)
+		}
+	}
+
+	for _, filename := range []string{ServerCert, ServerKey} {
+		if _, err = watcher.ensureCertificateFile(secret, filename); err != nil {
+			return false, err
+		}
+	}
+	isValid, err := watcher.ValidateCertificateExpiration(secret.Data[ServerCert], certificateRenewalInterval, time.Now())
+	if err != nil {
+		return false, err
+	} else if !isValid {
+		return false, fmt.Errorf("certificate is outdated")
+	}
+	return true, nil
+}
+
+func (watcher *CertificateWatcher) ensureCertificateFile(secret corev1.Secret, filename string) (bool, error) {
```

Comment:
i think here we also could remove the bool return value and just return err or nil

Reply:
if the bytes are not the same we should still fail somehow, but if you like it better I could rewrite it to return an adhoc error and remove the bool